# Descriptive and Inferential Statistics

### Population

A _population_ is a particular group of interest we want to study, such as “all seniors over the age of 65 in the North America,” “all golden retrievers in Scotland,” or “current high school sophomores at Los Altos High School.”

### Sample

A _sample_ is a subset of the population that is ideally random and unbiased, which we use to infer attributes about the population. We often have to study samples because polling the entire population is not always possible. 

### Bias

The way to overcome bias is to truly at random select students (in the example study) from the entire population, and they cannot elect themselves into or out of the sample voluntarily.

There are many types of bias, but they all have the same effect of distorting findings.

_Confirmation bias_ is gathering only data that supports your belief, which can even be done unknowingly. An example of this is following only social media accounts you politically agree with, reinforcing your beliefs rather than challenging them.

_Self-selection bias_ is when certain types of subjects are more likely to include themselves in the experiment. Walking onto a flight and polling the customers if they like the airline over other airlines, and using that to rank customer satisfaction among all airlines, is silly. Why? Many of those customers are likely repeat customers, and they have created self-selection bias.

_Survival bias_ captures only living and survived subjects, while the deceased ones are never accounted for.


## Descriptive Statistics

### Mean and Weighted Mean



In [2]:
# Number of pets each person owns
sample = [1, 3, 2, 5, 7, 0, 2, 3]

mean = sum(sample) / len(sample)
mean

2.875

The mean we commonly use (above) gives equal importance to each value. But we can manipulate the mean and give each item a different weight:

$$
\text{Weighted mean} = \frac{(x_1 \cdot w_1) + (x_2 \cdot w_2) + (x_3 \cdot w_3) + \ldots + (x_n \cdot w_n)}{w_1 + w_2 + w_3 + \ldots + w_n}
$$

This can be helpful when we want some values to contribute to the mean more than others. 

In [3]:
# Three exams of .20 weight each and final exam of .40 weight
sample = [90, 80, 63, 87]
weights = [.20, .20, .20, .40]

weighted_mean = sum(s * w for s,w in zip(sample, weights)) / sum(weights)
weighted_mean

81.4

### Median

The median is the middlemost value in a set of ordered values. You sequentially order the values, and the median will be the centermost value. If you have an even number of values, you average the two centermost values.

The median can be preferable in outlier-heavy situations (such as income-related data) over the mean, when your median is very different from your mean, that means you have a skewed dataset with outliers. 

In [6]:
# Number of pets each person owns
sample = [0, 1, 5, 7, 9, 10, 14]

def median(values):
    ordered = sorted(values)
    n = len(ordered)
    mid = int(n / 2) - 1 if n % 2 == 0 else int(n/2)

    if n % 2 == 0:
        return (ordered[mid] + ordered[mid+1]) / 2.0
    else:
        return ordered[mid]

median(sample)

7

### Mode

The mode is the most frequently occurring set of values. It primarily becomes useful when your data is repetitive and you want to find which values occur the most frequently.

When no value occurs more than once, there is no mode. When two values occur with an equal amount of frequency, then the dataset is considered _bimodal_.

In practicality, the mode is not used a lot unless your data is repetitive. This is commonly encountered with integers, categories, and other discrete variables.

In [8]:
# Number of pets each person owns
from collections import defaultdict

sample = [1, 3, 2, 5, 7, 0, 2, 3]

def mode(values):
    counts = defaultdict(lambda: 0)

    for s in values:
        counts[s] += 1

    max_count = max(counts.values())
    modes = [v for v in set(values) if counts[v] == max_count]
    return modes

mode(sample)

[2, 3]

### Variance and Standard Deviation

The _variance_ is a measure of how spread out our data is.

#### Population Variance and Standard Deviation

$$
\text{Population variance} = \frac{(x_1 - mean)^2 + (x_2 - mean)^2 + \ldots + (x_n - mean)^2}{N}
$$

More formally:

$$
\sigma^2 = \frac{\sum(x_i - \mu)^2}{N}
$$

In [15]:
# Number of pets each person owns
data = [0, 1, 5, 7, 9, 10, 14]

def variance(values):
    mean = sum(values) / len(values)
    _variance = sum((v - mean) ** 2 for v in values) / len(values)
    return _variance

variance(data)

21.387755102040813

So the variance for number of pets owned by my office staff is 21.387755. OK, but what does it exactly mean?

This number is larger than any of our observations because we did a lot squaring and summing, putting it on an entirely different metric. So how do we squeeze it back down so it’s back on the scale we started with?

Let’s take the square root of the variance, which gives us the _standard deviation_.

This is the variance scaled into a number expressed in terms of “number of pets,” which makes it a bit more meaningful:

$$
\sigma = \sqrt{\frac{\sum(x_i - \mu)^2}{N}}
$$

In [16]:
from math import sqrt

def std_dev(values):
    return sqrt(variance(values))

std_dev(data)

4.624689730353898

#### Sample Variance and Standard Deviation

There is an important tweak we need to apply to the two formulas above when we calculate for a sample:

$$
s^2 = \frac{\sum{(x_i - \overline{x})^1}}{n - 1}
$$

$$
s = \sqrt{\frac{\sum{(x_i - \overline{x})^1}}{n - 1}}
$$

When we average the squared differences, we divide by n–1 rather than the total number of items *n*.

Why would we do this? We do this to decrease any bias in a sample and not underestimate the variance of the population based on our sample. 

By counting values short of one item in our divisor, we increase the variance and therefore capture greater uncertainty in our sample.

In [2]:
from math import sqrt

# Number of pets each person owns
data = [0, 1, 5, 7, 9, 10, 14]


def variance(values, is_sample: bool = False):
    mean = sum(values) / len(values)
    _variance = sum((v - mean) ** 2 for v in values) / (len(values) - (1 if is_sample else 0))

    return _variance


def std_dev(values, is_sample: bool = False):
    return sqrt(variance(values, is_sample))

print("VARIANCE = {}".format(variance(data, is_sample=True))) # 24.95238095238095
print("STD DEV = {}".format(std_dev(data, is_sample=True))) # 4

VARIANCE = 24.95238095238095
STD DEV = 4.99523582550223


## The Normal Distribution

The *normal distribution*, also known as the *Gaussian distribution*, is a symmetrical bell-shaped distribution that has most mass around the mean, and its spread is defined as a standard deviation. The “tails” on either side become thinner as you move away from the mean.


**Properties of a Normal Distribution**

The normal distribution has several important properties that make it useful:

- It’s symmetrical; both sides are identically mirrored at the mean, which is the center.
- Most mass is at the center around the mean.
- It has a spread (being narrow or wide) that is specified by standard deviation.
- The “tails” are the least likely outcomes and approach zero infinitely but never touch zero.
- It resembles a lot of phenomena in nature and daily life, and even generalizes nonnormal problems because of the central limit theorem, which we will talk about shortly.

### The Probability Density Function (PDF)

$$
f(x) = \frac{1}{\sigma} \cdot \sqrt{2\pi} \cdot e^{-\frac{1}{2}(\frac{x-\mu^2}{\sigma})}
$$

The normal distribution is continuous. This means to retrieve a probability we need to integrate a range of x values to find an area.

In [5]:
# normal distribution, returns likelihood
def normal_pdf(x: float, mean: float, std_dev: float) -> float:
    return (1.0 / (2.0 * math.pi * std_dev ** 2) ** 0.5)  * math.exp(-1.0 * ((x - mean) ** 2 / (2.0 * std_dev ** 2)))

### The Cumulative Distribution Function (CDF)

With the normal distribution, the vertical axis is not the probability but rather the likelihood for the data. To find the probability we need to look at a given range, and then find the area under the curve for that range.

In [7]:
from scipy.stats import norm

mean = 64.43
std_dev = 2.99

norm.cdf(64.43, mean, std_dev)

0.5

We can deductively find the area for a middle range by subtracting areas. If we wanted to find the probability of observing a golden retriever between 62 and 66 pounds, we would calculate the area up to 66 and subtract the area up to 62:

In [9]:
from scipy.stats import norm

mean = 64.43
std_dev = 2.99

norm.cdf(66, mean, std_dev) - norm.cdf(62, mean, std_dev)

0.4920450147062894

### The Inverse CDF

We will encounter situations where we need to look up an area on the CDF and then return the corresponding x-value. Of course this is a backward usage of the CDF, so we will need to use the inverse CDF, which flips the axes.

For example, I want to find the weight that 95% of golden retrievers fall under:

In [10]:
from scipy.stats import norm

norm.ppf(.95, loc=64.43, scale=2.99)

69.3481123445849

I find that 95% of golden retrievers are 69.348 or fewer pounds.

You can also use the inverse CDF to generate random numbers that follow the normal distribution. If I want to create a simulation that generates one thousand realistic golden retriever weights, I just generate a random value between 0.0 and 1.0, pass it to the inverse CDF, and return the weight value:

In [11]:
import random
from scipy.stats import norm

for i in range(0,1000):
    random_p = random.uniform(0.0, 1.0)
    random_weight = norm.ppf(random_p,  loc=64.43, scale=2.99)
    print(random_weight)

62.35104011438862
65.01549007737131
69.78130759063522
60.707825174093564
62.892949890073716
62.92764598785052
65.53735758595354
59.91597654129303
67.79217412796397
71.02228798401022
64.08873490419425
63.5505704860073
68.65375341397953
63.99460141459659
65.57452113227616
62.5907810489494
64.28389361898734
66.96360206692466
62.128522754380114
68.63048537874948
61.32480719441721
62.48299765295826
71.11456347569506
65.13585019590134
66.2427900716726
62.55569754665831
60.25427526916987
58.62287196122942
61.41678706411152
65.76788589593839
60.552629185186944
66.24264592443397
61.29149202011201
63.546974696520216
63.60068060527812
61.71575203483936
61.4747167892669
70.38771567254955
61.60628979921501
65.21526848947981
66.49743102750041
65.59719244214676
66.93046034494796
65.15121634690651
68.4183698442515
65.23585665574144
66.42896936289269
63.705880794216974
67.72349460650207
58.15671012919269
63.964748918118076
63.48923989545005
65.7791944845353
62.05997163412638
65.49954375098442
62.503680

### Z-Scores

It is common to rescale a normal distribution so that the mean is 0 and the standard deviation is 1, which is known as the standard normal distribution. This makes it easy to compare the spread of one normal distribution to another normal distribution, even if they have different means and variances.

Of particular importance with the standard normal distribution is it expresses all x-values in terms of standard deviations, known as *Z-scores*. Turning an x-value into a Z-score uses a basic scaling formula:

$$
z = \frac{x - \mu}{\sigma}
$$

Example:

We have two homes from two different neighborhoods. Neighborhood A has a mean home value of $140,000 and standard deviation of $3,000. Neighborhood B has a mean home value of $800,000 and standard deviation of $10,000.

$$
\mu_A = 140,000
\\\\
\mu_B = 800,000
\\\\
\sigma_A = 3,000
\\\\
\sigma_B = 10,000
$$

Now we have two homes from each neighborhood. House A from neighborhood A is worth $150,000 and house B from neighborhood B is worth $815,000. Which home is more expensive relative to the average home in its neighborhood?

$$
x_A = 150,000
\\\\
x_B = 815,000
$$

Using $z = \frac{x - \mu}{\sigma}$:

$$
z_A = \frac{150000 - 140000}{3000} = 3.\overline{333}
\\\\
z_B = \frac{800000 - 140000}{10000} = 1.5
$$

So the house in neighborhood A is actually much more expensive relative to its neighborhood than the house in neighborhood B, as they have Z-scores of $3.\overline{333}$ and $1.5$, respectively.

In [13]:
# Turn Z-scores into x-values and vice versa

def z_score(x, mean, std):
    return (x - mean) / std


def z_to_x(z, mean, std):
    return (z * std) + mean


mean = 140000
std_dev = 3000
x = 150000

# Convert to Z-score and then back to X
z = z_score(x, mean, std_dev)
back_to_x = z_to_x(z, mean, std_dev)

print("Z-Score: {}".format(z))  # Z-Score: 3.333
print("Back to X: {}".format(back_to_x))  # Back to X: 150000.0

Z-Score: 3.3333333333333335
Back to X: 150000.0


## Inferential Statistics